### PR TITLE
wip: reducing the times of invoking fs_stat in oldfiles

### PR DESCRIPTION
**Pull request by haolian9**
hi, while reading providers/oldfiles.lua, i begin to wonder if we can reduce the times of invoking fs_stat.

and the very first question always is: is it a real problem? and it's worth it?
honestly i'm not sure, i am just having fun. well, maybe yes. the default length of v:oldfiles is 100, and these files might be stored in different disks: ssd, hdd, usb stick.

at the very begin, and just like [this commit](https://github.com/ibhagwan/fzf-lua/commit/9b418128d7a0ebc2a740b11938c69f75eabb9c3a), i thought we can cache the results of fs_stat.

as i dig deeper, i figured some other approaches, bases on several facts:
1. `:h 'oldfiles' -> `The List can be modified, but this has no effect on what is stored in the |shada| file later.`
2. `:h :oldfiles` -> `This list is read on startup and only changes afterwards with \`:rshada!\`.`

with `1`, we process v:oldfiles once, then store the results to v:oldfiles for later use, and add a special mark to it for checking if it has been changed by others. while this could not be acceptable to users.

with `2`, we process v:oldfiles once, and store it in providers.oldfiles module, and invalid it when `:rshada`. in which cmdlineleave and cmdwinleave events and `quote:` can be used to watch if :rshada is being ran.

thanks for your reading, i hope these texts have not wasted mush time from you, and them did have some valid points.

---

after all this, i would just delegate these concerns to end users and let them offer a source of provider.oldfiles, haha.

### CODE REVIEW

**Comment by ibhagwan** [comment]
Ty @haolian9.

> and the very first question always is: is it a real problem? and it's worth it?

This is the real question, IMHO, it’s never been a problem, certainly not for the default of 100 files.

I did test this with up to 100k entries, you can see the remnants of my debug prints before and after:
https://github.com/ibhagwan/fzf-lua/blob/b58799714513083bd646f0b716e29f224ebba48e/lua/fzf-lua/providers/oldfiles.lua#L53-L60

When originally improving performance of `oldfiles` in #343, I tested up to 1M entries, although slower, even 1M wasn’t an issue as fzf is being async fed (using the `coroutine`) and while I’m usually very open to performance enhancements this does come with the caveat of having to manage the cache as old files are never cleared, if someone decides to delete/rename files the cache needs to be updated which gets us back to square 1 (imagine someone “lives” inside neovim, they’re bound to make fs modifications).

Although probably not a huge issue I also dislike having unnecessary memory hogs inside the plug-in, which is why I went with a circular cache for the shell functions ID cache:

https://github.com/ibhagwan/fzf-lua/blob/b58799714513083bd646f0b716e29f224ebba48e/lua/fzf-lua/shell.lua#L9-L25

Bottom line, I would be against adding such a change as this has the potential of being a “voodoo bug” one day after I’ve forgotten we added this “forever cache”.

Much appreciate the effort!
